### PR TITLE
feat: add lru cache

### PR DESCRIPTION
**Pull request by TonPC64**
WIKI: https://en.wikipedia.org/wiki/Cache_replacement_policies#Least_recently_used_(LRU)

### CODE REVIEW

File: cache/lru/lru.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
+package lru
+
+import "container/list"
```
Comment (tjgurwara99) [inline cache/lru/lru.go:3]: Could you instead use the list implementation [here](https://github.com/TheAlgorithms/Go/blob/eaa2be2ffe9295b9eb996ec49af6e0e2e6381595/structure/linkedlist/doubly.go#L18)?

Comment (TonPC64) [reply cache/lru/lru.go:3]: I try to use it but it missing a lot of necessary function like `Remove` or `Move`

Comment (tjgurwara99) [reply cache/lru/lru.go:3]: Perfect opportunity for you to add that if you're interested 😄 If not let me know.

Comment (TonPC64) [reply cache/lru/lru.go]: @tjgurwara99 I Fixed it please review

File: cache/lru/lru.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
+package lru
```
Comment (tjgurwara99) [inline cache/lru/lru.go:1]: Move everything to `cache` package.

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]

File: cache/lru/lru.go (hunk 2 above)
Comment (TonPC64) [reply cache/lru/lru.go]: moved.

File: cache/lru_test.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
+package cache_test
+
+import (
+	"testing"
+
+	lru "github.com/TheAlgorithms/Go/cache"
```
Comment (tjgurwara99) [inline cache/lru_test.go:6]: This is your change. There is no big dependency on the name `lru` for you to add `lru` here. Using shortcuts is not a good practice - fixing is simply a version of `gofmt -r 'lru.a-> cache.a' -s -w path/to/this/file.go`

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]

File: cache/lru_test.go (hunk 3 above)
Comment (TonPC64) [reply cache/lru_test.go]: I fixed by rename function `New` to `NewLRU` for using by `cache.NewLRU()`

File: structure/linkedlist/doubly.go [hunk 4]
```diff
@@ -16,86 +16,99 @@ import "fmt"
 //
 // and call linkedlist.Doubly to create a new doubly linked list.
 type Doubly[T any] struct {
-	Head *Node[T]
+	Head Node[T]
```
Comment (tjgurwara99) [inline structure/linkedlist/doubly.go:19]: Shouldn't this be a pointer?

File: cache/lru_test.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+		key, value := "1", 1
+		cache.Put(key, value)
+		got := cache.Get(key)
+
+		if got != value {
+			t.Errorf("Put: %v, Got: %v", value, got)
+		}
+	})
+
+	t.Run("Test 2: Get data without cache should got nil", func(t *testing.T) {
```
Comment (tjgurwara99) [inline cache/lru_test.go:22]:
```suggestion
	t.Run("Test 2: Get data not stored in cache should return nil", func(t *testing.T) {
```

File: cache/lru_test.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+		got := cache.Get("2")
+		if got != nil {
+			t.Errorf("Got: %v", got)
+		}
+	})
+
+	t.Run("Test 3: Put data over capacity and Get should got nil", func(t *testing.T) {
```
Comment (tjgurwara99) [inline cache/lru_test.go:29]:
```suggestion
	t.Run("Test 3: Put data over capacity and Get should return nil", func(t *testing.T) {
```

File: cache/lru_test.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+		cache.Put("3", 3)
+		cache.Put("4", 4)
+
+		got := cache.Get("1")
+		if got != nil {
+			t.Errorf("Got: %v", got)
+		}
+	})
+
+	t.Run("test 4: Put key over capacity but key already existing", func(t *testing.T) {
```
Comment (tjgurwara99) [inline cache/lru_test.go:40]:
```suggestion
	t.Run("test 4: Put key over capacity but recent key exists", func(t *testing.T) {
```

File: cache/lru_test.go [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+func TestLRU(t *testing.T) {
+	cache := cache.NewLRU(3)
+
+	t.Run("Test 1: Put number and Get is correct", func(t *testing.T) {
+		key, value := "1", 1
+		cache.Put(key, value)
+		got := cache.Get(key)
+
+		if got != value {
+			t.Errorf("Put: %v, Got: %v", value, got)
```
Comment (tjgurwara99) [inline cache/lru_test.go:18]:
```suggestion
			t.Errorf("expected: %v, got: %v", value, got)
```

File: cache/lru_test.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+		}
+	})
+
+	t.Run("Test 2: Get data without cache should got nil", func(t *testing.T) {
+		got := cache.Get("2")
+		if got != nil {
+			t.Errorf("Got: %v", got)
```
Comment (tjgurwara99) [inline cache/lru_test.go:25]:
```suggestion
			t.Errorf("expected: nil, got: %v", got)
```

File: cache/lru_test.go [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+	})
+
+	t.Run("Test 3: Put data over capacity and Get should got nil", func(t *testing.T) {
+		cache.Put("2", 2)
+		cache.Put("3", 3)
+		cache.Put("4", 4)
+
+		got := cache.Get("1")
+		if got != nil {
+			t.Errorf("Got: %v", got)
```
Comment (tjgurwara99) [inline cache/lru_test.go:36]:
```suggestion
			t.Errorf("expected: nil, got: %v", got)
```

File: cache/lru_test.go [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+	t.Run("test 4: Put key over capacity but key already existing", func(t *testing.T) {
+		cache.Put("4", 4)
+		cache.Put("3", 3)
+		cache.Put("2", 2)
+		cache.Put("1", 1)
+
+		got := cache.Get("4")
+		if got != nil {
+			t.Errorf("Got: %v", got)
```
Comment (tjgurwara99) [inline cache/lru_test.go:48]:
```suggestion
			t.Errorf("expected: nil, got: %v", got)
```

File: cache/lru_test.go [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+		}
+
+		expected := 3
+		got = cache.Get("3")
+
+		if got != expected {
+			t.Errorf("Expected: %v, Got: %v", expected, got)
```
Comment (tjgurwara99) [inline cache/lru_test.go:55]:
```suggestion
			t.Errorf("expected: %v, got: %v", expected, got)
```

**Review by tjgurwara99 — CHANGES_REQUESTED** [review]

**Review by tjgurwara99 — APPROVED** [review]

**Review by raklaptudirm — APPROVED** [review]

**Pull request merged**